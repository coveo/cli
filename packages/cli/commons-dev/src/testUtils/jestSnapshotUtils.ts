// See https://github.com/oclif/core/blob/f996086397158b5beafc4254718c78fd93d23b1b/src/errors/errors/cli.ts#L55-L60
export const formatCliLog = (input: string) =>
  input.replaceAll(/\r\n/gm, '\n').replaceAll(/»/gm, '›');

export const formatAbsolutePath = (input: string) =>
  input.replaceAll(/\w:\\{2}/gm, '/').replaceAll(/\\{2}/gm, '/');

export const removeDotCmd = (input: string) => input.replaceAll(/\.cmd/gm, '');
