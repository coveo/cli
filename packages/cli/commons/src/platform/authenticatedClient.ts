import {setGlobalDispatcher, ProxyAgent, FormData, fetch} from 'undici';
import PlatformClient from '@coveo/platform-client';
import {Config, Configuration} from '../config/config';
import {castEnvironmentToPlatformClient} from './environment';
import globalConfig from '../config/globalConfig';
import {lt} from 'semver';

export class AuthenticatedClient {
  public cfg: Config;
  public constructor() {
    this.cfg = new Config(globalConfig.get().configDir);
  }

  public async isLoggedIn() {
    const {accessToken} = this.cfg.get();
    return accessToken !== undefined;
  }

  public async isExpired() {
    try {
      const c = await this.getClient();
      await c.initialize();
      return false;
    } catch (e) {
      return true;
    }
  }

  public async getClient(overrideConfig?: Partial<Configuration>) {
    const configFromDisk = this.cfg.get();
    const resolvedConfig = {...configFromDisk, ...overrideConfig};
    const globalRequestSettings: Record<string, unknown> = {};
    const proxyServer =
      process.env['https_proxy'] || process.env['HTTPS_PROXY'];
    if (proxyServer) {
      setGlobalDispatcher(new ProxyAgent(proxyServer));
    }
    if (lt(process.version, '18.0.0')) {
      Object.assign(global, {FormData, fetch});
    }
    return new PlatformClient({
      globalRequestSettings,
      environment: castEnvironmentToPlatformClient(resolvedConfig.environment),
      region: resolvedConfig.region,
      organizationId: resolvedConfig.organization,
      accessToken: resolvedConfig.accessToken!,
    });
  }

  public async getAllOrgsUserHasAccessTo() {
    const platformClient = await this.getClient();
    return platformClient.organization.list();
  }

  public async getUserHasAccessToOrg(org: string) {
    const orgs = await this.getAllOrgsUserHasAccessTo();
    return orgs.some((o) => o.id === org);
  }

  public async createImpersonateApiKey(name: string) {
    const platformClient = await this.getClient();
    return await platformClient.apiKey.create({
      displayName: `cli-${name}`,
      description: 'Generated by the Coveo CLI',
      enabled: true,
      privileges: [
        {targetDomain: 'IMPERSONATE', targetId: '*', owner: 'SEARCH_API'},
      ],
    });
  }

  public async getUsername() {
    const {anonymous, organization} = this.cfg.get();
    if (anonymous) {
      return `api-key@${organization}`;
    }
    const authenticatedClient = new AuthenticatedClient();
    const platformClient = await authenticatedClient.getClient();
    await platformClient.initialize();

    return (await platformClient.user.get()).providerUsername;
  }
}

export enum AuthenticationStatus {
  LOGGED_IN,
  EXPIRED,
  LOGGED_OUT,
}

export async function getAuthenticationStatus() {
  const authenticatedClient = new AuthenticatedClient();

  if (!(await authenticatedClient.isLoggedIn())) {
    return AuthenticationStatus.LOGGED_OUT;
  }

  if (await authenticatedClient.isExpired()) {
    return AuthenticationStatus.EXPIRED;
  }

  return AuthenticationStatus.LOGGED_IN;
}
