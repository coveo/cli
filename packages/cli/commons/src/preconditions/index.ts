export * from './authenticated';
export * from './apiKeyPrivilege';
export * from './preconditions';
