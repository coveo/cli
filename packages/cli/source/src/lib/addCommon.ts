import {CliUx} from '@oclif/core';
import {red, dim, green} from 'chalk';
import {BuiltInTransformers, errors} from '@coveo/push-api-client';
import {withNormalizeInvalidFields} from './commonFlags';
import {pluralizeIfNeeded} from '@coveo/cli-commons/utils/string';
import dedent from 'ts-dedent';

const allowedCharRegExp = /^[a-z]+[a-z0-9_]*$/;
interface PrintableUnsupportedField {
  original: string;
  normalized: string;
  valid: boolean;
}

export const formatErrorMessage = (err: unknown) => {
  const {UnsupportedFieldError} = errors;
  if (err instanceof UnsupportedFieldError) {
    const normalizations = getNormalizations(err.unsupportedFields);
    const {fixables, unfixables} = colorizeFields(normalizations);
    const isFixable = unfixables.length === 0;

    printInvalidFieldTable(fixables);
    printInvalidFieldTable(unfixables);
    err.message = getFormattedMessage(isFixable);
  }
};

const getFormattedMessage = (isFixable: boolean) => {
  const normalizeFlag = Object.keys(withNormalizeInvalidFields())[0];
  let message = `Run the same command using the \`--${normalizeFlag}\` flag to automatically normalize field names while pushing your data.`;
  if (!isFixable) {
    message = dedent`Cannot normalize some of the invalid field names detected in your data
      Review your data and ensure all field names as shown in the second table have at least one alphabetic character.`;
  }
  return message;
};

const colorizeFields = (fields: PrintableUnsupportedField[]) => {
  const fixables: PrintableUnsupportedField[] = [];
  const unfixables: PrintableUnsupportedField[] = [];
  for (const field of fields) {
    if (field.valid) {
      const normalized = green(field.normalized);
      fixables.push({...field, normalized});
    } else {
      const normalized =
        field.normalized === ''
          ? dim('(empty field name)')
          : red(field.normalized);
      unfixables.push({...field, normalized});
    }
  }
  return {fixables, unfixables};
};

const getNormalizations = (
  unsupportedFields: [string, string][]
): PrintableUnsupportedField[] => {
  return unsupportedFields.map(([original, transformed]) => {
    const normalized = BuiltInTransformers.toLowerCase(transformed);
    return {
      original,
      normalized,
      valid: isFieldNameValid(normalized),
    };
  });
};

const printInvalidFieldTable = (fields: {valid: boolean}[]) => {
  const count = fields.length;
  const pluralized = pluralizeIfNeeded(['field', 'fields'], count);
  if (count === 0) {
    return;
  }

  const fixable = fields[0].valid;
  CliUx.ux.log(
    ` ${count} ${pluralized} detected in your data can${
      fixable ? '' : 'not'
    } be normalized`
  );
  CliUx.ux.table(fields, {
    original: {header: 'Original'},
    normalized: {header: 'Normalized'},
  });
  logNewLine();
};

export const logNewLine = (lines = 1) => {
  for (let i = 0; i < lines; i++) {
    CliUx.ux.log('');
  }
};

const isFieldNameValid = (fieldName: string): boolean => {
  return allowedCharRegExp.test(fieldName.toLowerCase());
};
