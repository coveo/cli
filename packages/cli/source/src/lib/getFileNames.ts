import {CliUx} from '@oclif/core';

export interface FilesOrFolders {
  file?: string[]; // TODO: CDX-856: remove flag
  folder?: string[]; // TODO: CDX-856: remove flag
  files?: string[];
}

export function getFileNames(entries: FilesOrFolders, commandName: string) {
  // TODO: CDX-856: only read --files flag
  const entryNames = [
    ...(entries.file ?? []),
    ...(entries.folder ?? []),
    ...(entries.files ?? []),
  ];

  if (entryNames.length === 0) {
    CliUx.ux.error(
      `You must set the \`files\` flag. Use \`${commandName} --help\` to get more information.`
    );
  }

  return entryNames;
}
