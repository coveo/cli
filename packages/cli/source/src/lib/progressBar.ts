import {MultiBar, SingleBar} from 'cli-progress';
import {cyan} from 'chalk';
import {CliUx} from '@oclif/core';

export class ProgressBar {
  private bar?: SingleBar;
  private multibar: MultiBar;
  private static container = {
    format: 'Progress | {bar} | {value}/{total} documents',
    hideCursor: true,
    // barCompleteChar: '\u2588',
    // barIncompleteChar: '\u2591',

    // only the bars will be cleared, not the logged content
    clearOnComplete: true,
    stopOnComplete: true,

    // important! redraw everything to avoid "empty" completed bars
    forceRedraw: true,
  };

  public constructor() {
    this.multibar = this.createBarContainer();
  }

  public increment(count: number) {
    this.bar?.increment(count);
    return this;
  }

  public ensureInitialization(startValue: number, total?: number) {
    if (total && !this.bar) {
      CliUx.ux.action.stop('in progress');
      this.bar = this.multibar.create(total, startValue);
    }
    return this;
  }

  public log(message: string) {
    // linebreak is required to not break the progress bar
    this.multibar.log(`${message}\n`);
  }

  private createBarContainer() {
    return new MultiBar(ProgressBar.container);
  }
}
