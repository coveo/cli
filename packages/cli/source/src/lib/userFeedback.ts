import {Command} from '@oclif/core';
import {green} from 'chalk';
import {validate} from 'jsonschema';
import dedent from 'ts-dedent';
import {
  APIError,
  AxiosErrorFromAPI,
  AxiosErrorFromAPISchema,
} from '@coveo/cli-commons/errors/apiError';
import {UnknownError} from '@coveo/cli-commons/errors/unknownError';

export interface AxiosResponse {
  status: number;
  statusText: string;
}

export const successMessage = (
  cmd: Command,
  tagLine: string,
  res?: AxiosResponse
) => {
  let message = dedent(`
      ${tagLine}
      `);
  if (res) {
    message += `Status code: ${green(res.status, res.statusText)}
    `;
  }
  cmd.log(message);
};

export const errorMessage = (
  cmd: Command,
  tagLine: string,
  e: unknown,
  options = {exit: false}
) => {
  const error = isErrorFromAPI(e)
    ? new APIError(e, tagLine)
    : new UnknownError(e);

  if (options.exit) {
    throw error;
  } else {
    cmd.warn(error.message);
  }
};

function isErrorFromAPI(error: unknown): error is AxiosErrorFromAPI {
  return validate(error, AxiosErrorFromAPISchema).valid;
}
