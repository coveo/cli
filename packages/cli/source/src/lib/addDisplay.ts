import {green} from 'chalk';
import {UploadBatchCallbackData} from '@coveo/push-api-client';
import {errorMessage, successMessage} from './userFeedback';
import {Plurable, pluralizeIfNeeded} from '@coveo/cli-commons/utils/string';
import {CliUx} from '@oclif/core';
import {UploadProgress} from '@coveo/push-api-client/dist/definitions/interfaces';
import {logNewLine} from './addCommon';

export interface AddSummary {
  added: number;
  failed: number;
  remaining?: number;
  total?: number;
}

export class AddDisplay {
  private summary: AddSummary = {
    added: 0,
    failed: 0,
  };

  public successMessageOnAdd({
    batch,
    files,
    res,
    progress,
  }: UploadBatchCallbackData) {
    // Display the first 5 files (from the list of all files) being processed for end user feedback.
    // Don't want to clutter the output too much if the list is very long.
    let fileNames = files.slice(0, 5).join(', ');
    if (files.length > 5) {
      fileNames += ` and ${files.length - 5} more ...`;
    }
    const numAdded = batch.length;
    const plurableDoc: Plurable = ['document', 'documents'];
    successMessage(
      `Success: ${green(numAdded)} ${pluralizeIfNeeded(
        plurableDoc,
        numAdded
      )} accepted by the API from ${green(fileNames)}.`,
      res
    );

    this.refreshSummary(batch.length, 0, progress);
    // TODO: CDX-712 display progress bar if progress data is available
  }

  public errorMessageOnAdd(
    e: unknown,
    {batch, progress}: UploadBatchCallbackData
  ) {
    this.refreshSummary(0, batch.length, progress);
    errorMessage('Error while trying to add document.', e, {exit: true});
  }

  private refreshSummary(
    added: number,
    failed: number,
    progress?: UploadProgress
  ) {
    this.summary.added += added;
    this.summary.failed += failed;
    this.summary.remaining = progress?.remainingDocumentCount;
    this.summary.total = progress?.totalDocumentCount;
  }

  public printSummary() {
    const {added, failed, remaining, total} = this.summary;
    const pluralized = (docCount: number) =>
      pluralizeIfNeeded(['document', 'documents'], docCount);

    const formatCount = (count: number) => {
      return `${count}${total ? ['/', total].join('') : ''}`;
    };

    CliUx.ux.styledHeader('Push Summary');

    CliUx.ux.log(
      `${formatCount(added)} ${pluralized(added)} successfully sent to the API`
    );

    if (failed > 0) {
      CliUx.ux.log(
        `${formatCount(failed)} ${pluralized(failed)} failed to be sent`
      );
    }

    if (remaining && remaining > 0) {
      CliUx.ux.log(
        `${formatCount(remaining)} unprocessed ${pluralized(remaining)}`
      );
    }

    logNewLine();
  }
}
