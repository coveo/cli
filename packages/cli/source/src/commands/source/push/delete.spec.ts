jest.mock('@coveo/cli-commons/preconditions/trackable');
jest.mock('@coveo/cli-commons/preconditions/authenticated');

jest.mock('@coveo/cli-commons/platform/authenticatedClient');
jest.mock('@coveo/push-api-client', () => {
  const mock = jest.createMockFromModule<object>('@coveo/push-api-client');
  const real = jest.requireActual('@coveo/push-api-client');
  return {
    __esModule: true,
    ...mock,
    errors: real.errors,
  };
});

jest.mock('../../../lib/userFeedback');

import {test} from '@oclif/test';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {errors, PushSource} from '@coveo/push-api-client';
import {IsAuthenticated} from '@coveo/cli-commons/preconditions';
import {mockPreconditions} from '@coveo/cli-commons/preconditions/mockPreconditions';
import {formatCliLog} from '@coveo/cli-commons-dev/testUtils/jestSnapshotUtils';
import {errorMessage, successMessage} from '../../../lib/userFeedback';
import {Response} from 'undici';

describe('source:push:delete', () => {
  const mockedIsAuthenticated = jest.mocked(IsAuthenticated);
  const mockedClient = jest.mocked(AuthenticatedClient);
  const mockedSource = jest.mocked(PushSource);
  const mockDeleteOlderThan = jest.fn();
  const mockedErrorMessage = jest.mocked(errorMessage);
  const mockedSuccessMessage = jest.mocked(successMessage);
  const mockDeleteDocument = jest.fn();

  const doMockPreconditions = function () {
    const preconditionStatus = {
      authentication: true,
    };
    const mockedPreconditions = mockPreconditions(preconditionStatus);
    mockedIsAuthenticated.mockReturnValue(mockedPreconditions.authentication);
  };

  const doMockAuthenticatedClient = () => {
    mockedClient.mockImplementation(
      () =>
        ({
          cfg: {
            get: () => ({
              accessToken: 'the_token',
              organization: 'the_org',
              environment: 'some_env',
              region: 'MARS',
            }),
          },
        } as unknown as AuthenticatedClient)
    );
  };

  const doMockSourceDelete = () => {
    mockedSource.mockImplementation(
      () =>
        ({
          deleteDocumentsOlderThan: mockDeleteOlderThan.mockReturnValue(
            Promise.resolve({
              status: 202,
              title: 'TIGUIDOU',
              detail: 'all is well',
            })
          ),
          deleteDocument: mockDeleteDocument.mockReturnValue(
            Promise.resolve({
              status: 202,
              title: 'TIGUIDOU',
              detail: 'right trou',
            })
          ),
        } as unknown as PushSource)
    );
  };

  beforeEach(() => {
    doMockAuthenticatedClient();
    doMockPreconditions();
    doMockSourceDelete();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test
    .stdout()
    .stderr()
    .command(['source:push:delete', 'mysource'])
    .catch((err) => {
      expect(err.message).toMatchSnapshot();
    })
    .it('throws when no flags are specified');

  test
    .stdout()
    .stderr()
    .command(['source:push:delete', 'mysource', '-x', 'foo', '-d', 'bar'])
    .catch((err) => {
      expect(err.message).toMatchSnapshot();
    })
    .it(
      'throws when incompatible flags for olderThan and documentUri are passed'
    );

  test
    .stdout()
    .stderr()
    .command(['source:push:delete', 'mysource', '-d', '2000/01/01'])
    .it('pass correct configuration information to push-api-client', () => {
      expect(mockedSource).toHaveBeenCalledWith('the_token', 'the_org', {
        environment: 'some_env',
        region: 'MARS',
      });
    });

  [
    '2000/01/01',
    '2000-01-01',
    '2000-01-01T06:00:00+00:00',
    'Monday, 2000-Jan-01 06:00:00 UTC',
  ].forEach((testCase) => {
    test
      .stdout()
      .stderr()
      .command(['source:push:delete', 'mysource', '-d', testCase])
      .it(
        `pass correct values to push-api-client when deleting with date as string: ${testCase}`,
        () => {
          expect(mockDeleteOlderThan).toHaveBeenCalledWith(
            'mysource',
            testCase
          );
        }
      );
  });

  test
    .stdout()
    .stderr()
    .command(['source:push:delete', 'mysource', '-d', '123123123'])
    .it(
      'pass correct values to push-api-client when deleting with date as number',
      () => {
        expect(mockDeleteOlderThan).toHaveBeenCalledWith('mysource', 123123123);
      }
    );

  test
    .do(() => {
      mockDeleteOlderThan.mockReturnValueOnce({
        status: 999,
        title: 'WOOSH',
        detail: 'this document is gone',
      });
    })
    .stdout()
    .stderr()
    .command(['source:push:delete', 'mysource', '-d', '12345'])
    .it(
      'returns an information message on successful deletion with older than',
      () => {
        expect(mockedSuccessMessage.mock.lastCall).toMatchSnapshot();
      }
    );

  test
    .do(async () => {
      const err = await errors.FetchError.build({
        json: () =>
          Promise.resolve({
            status: 412,
            title: 'BAD_REQUEST',
            detail: 'this is a bad request and you should feel bad',
          }),
      } as Response);
      mockDeleteOlderThan.mockRejectedValueOnce(err).mockRejectedValueOnce(err);
    })
    .stdout()
    .stderr()
    .command(['source:push:delete', 'mysource', '-d', '12345'])
    .it(
      'returns an information message on deletion failure with older than',
      () => {
        expect(mockedErrorMessage.mock.lastCall).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      mockDeleteDocument.mockReturnValueOnce({
        status: 999,
        title: 'WOOSH',
        detail: 'this document is gone',
      });
    })
    .stdout()
    .stderr()
    .command(['source:push:delete', 'mysource', '-x', 'https://foo.com'])
    .it(
      'returns an information message on successful deletion with document uri',
      () => {
        expect(mockedSuccessMessage.mock.lastCall).toMatchSnapshot();
      }
    );

  test
    .stdout()
    .stderr()
    .command([
      'source:push:delete',
      'mysource',
      '-x',
      'https://foo.com',
      '-x',
      'https://foo.com/2',
    ])
    .it(
      'returns an information message on successful deletion with multiple document uri',
      () => {
        expect(mockedSuccessMessage.mock.lastCall).toMatchSnapshot();
      }
    );

  test
    .do(async () => {
      const err = await errors.FetchError.build({
        json: () =>
          Promise.resolve({
            status: 412,
            title: 'BAD_REQUEST',
            detail: 'this is a bad request and you should feel bad',
          }),
      } as Response);
      mockDeleteDocument.mockRejectedValueOnce(err);
    })
    .stdout()
    .stderr()
    .command(['source:push:delete', 'mysource', '-x', 'https://foo.com'])
    .it(
      'returns an information message on deletion failure with document uri',
      () => {
        expect(mockedErrorMessage.mock.lastCall).toMatchSnapshot();
      }
    );

  test
    .do(async () => {
      const err = await errors.FetchError.build({
        json: () =>
          Promise.resolve({
            status: 412,
            title: 'BAD_REQUEST',
            detail: 'this is a bad request and you should feel bad',
          }),
      } as Response);
      mockDeleteDocument.mockRejectedValueOnce(err).mockRejectedValueOnce(err);
    })
    .stdout()
    .stderr()
    .command([
      'source:push:delete',
      'mysource',
      '-x',
      'https://foo.com',
      '-x',
      'https://foo.com/2',
    ])
    .it(
      'returns an information message on deletion failure with multiple document uri',
      () => {
        expect(mockedErrorMessage.mock.lastCall).toMatchSnapshot();
      }
    );

  test
    .do(async () => {
      const err = await errors.FetchError.build({
        json: () =>
          Promise.resolve({
            status: 412,
            title: 'BAD_REQUEST',
            detail: 'this is a bad request and you should feel bad',
          }),
      } as Response);
      mockDeleteDocument.mockRejectedValueOnce(err);
    })
    .stdout()
    .stderr()
    .command([
      'source:push:delete',
      'mysource',
      '-x',
      'https://foo.com',
      '-x',
      'https://foo.com/2',
    ])
    .it(
      'returns an information message on deletion and success failure with multiple document uri',
      () => {
        expect(mockedErrorMessage.mock.lastCall).toMatchSnapshot();
      }
    );

  test
    .stdout()
    .stderr()
    .command([
      'source:push:delete',
      'mysource',
      ...Array(21)
        .fill(null)
        .flatMap((_, index) => ['-x', `https://foo.com/${index}`]),
    ])
    .it(
      'should warn the user when he tries to delete too many items and stop there',
      (ctx) => {
        expect(mockDeleteDocument).not.toBeCalled();
        expect(formatCliLog(ctx.stderr)).toMatchSnapshot();
      }
    );
});
