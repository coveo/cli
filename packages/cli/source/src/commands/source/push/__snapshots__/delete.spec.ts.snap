// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`source:push:delete returns an information message on deletion and success failure with multiple document uri 1`] = `
"The delete request for document: https://foo.com/2 was accepted by the Push API.Status code: 202 right trou
    
  console.error
     â€º   Warning: Error while trying to delete document: https://foo.com.
     â€º
     â€º   Status code: 412
     â€º   Error code: BAD_REQUEST
     â€º   Message: this is a bad request and you should feel bad

      32 |     throw error;
      33 |   } else {
    > 34 |     CliUx.ux.warn(error.message);
         |              ^
      35 |   }
      36 | };
      37 |

      at Object.warn (../../../node_modules/@oclif/core/lib/errors/index.js:59:13)
      at warn (src/lib/userFeedback.ts:34:14)
      at SourcePushDelete.errorMessageOnDeletion (src/commands/source/push/delete.ts:126:12)
      at errorMessageOnDeletion (src/commands/source/push/delete.ts:119:16)
          at async Promise.all (index 0)
      at SourcePushDelete.run (src/commands/source/push/delete.ts:82:7)
      at SourcePushDelete._run (../../../node_modules/@oclif/core/lib/command.js:81:22)
      at Config.runCommand (../../../node_modules/@oclif/core/lib/config/config.js:272:25)
      at Object.run (../../../node_modules/@oclif/test/lib/command.js:21:13)
      at Object.run (../../../node_modules/fancy-test/lib/base.js:68:25)

"
`;

exports[`source:push:delete returns an information message on deletion failure with document uri 1`] = `
"  console.error
     â€º   Warning: Error while trying to delete document: https://foo.com.
     â€º
     â€º   Status code: 412
     â€º   Error code: BAD_REQUEST
     â€º   Message: this is a bad request and you should feel bad

      32 |     throw error;
      33 |   } else {
    > 34 |     CliUx.ux.warn(error.message);
         |              ^
      35 |   }
      36 | };
      37 |

      at Object.warn (../../../node_modules/@oclif/core/lib/errors/index.js:59:13)
      at warn (src/lib/userFeedback.ts:34:14)
      at SourcePushDelete.errorMessageOnDeletion (src/commands/source/push/delete.ts:126:12)
      at errorMessageOnDeletion (src/commands/source/push/delete.ts:119:16)
          at async Promise.all (index 0)
      at SourcePushDelete.run (src/commands/source/push/delete.ts:82:7)
      at SourcePushDelete._run (../../../node_modules/@oclif/core/lib/command.js:81:22)
      at Config.runCommand (../../../node_modules/@oclif/core/lib/config/config.js:272:25)
      at Object.run (../../../node_modules/@oclif/test/lib/command.js:21:13)
      at Object.run (../../../node_modules/fancy-test/lib/base.js:68:25)

"
`;

exports[`source:push:delete returns an information message on deletion failure with multiple document uri 1`] = `
"  console.error
     â€º   Warning: Error while trying to delete document: https://foo.com.
     â€º
     â€º   Status code: 412
     â€º   Error code: BAD_REQUEST
     â€º   Message: this is a bad request and you should feel bad

      32 |     throw error;
      33 |   } else {
    > 34 |     CliUx.ux.warn(error.message);
         |              ^
      35 |   }
      36 | };
      37 |

      at Object.warn (../../../node_modules/@oclif/core/lib/errors/index.js:59:13)
      at warn (src/lib/userFeedback.ts:34:14)
      at SourcePushDelete.errorMessageOnDeletion (src/commands/source/push/delete.ts:126:12)
      at errorMessageOnDeletion (src/commands/source/push/delete.ts:119:16)
          at async Promise.all (index 0)
      at SourcePushDelete.run (src/commands/source/push/delete.ts:82:7)
      at SourcePushDelete._run (../../../node_modules/@oclif/core/lib/command.js:81:22)
      at Config.runCommand (../../../node_modules/@oclif/core/lib/config/config.js:272:25)
      at Object.run (../../../node_modules/@oclif/test/lib/command.js:21:13)
      at Object.run (../../../node_modules/fancy-test/lib/base.js:68:25)

  console.error
     â€º   Warning: Error while trying to delete document: https://foo.com/2.
     â€º
     â€º   Status code: 412
     â€º   Error code: BAD_REQUEST
     â€º   Message: this is a bad request and you should feel bad

      32 |     throw error;
      33 |   } else {
    > 34 |     CliUx.ux.warn(error.message);
         |              ^
      35 |   }
      36 | };
      37 |

      at Object.warn (../../../node_modules/@oclif/core/lib/errors/index.js:59:13)
      at warn (src/lib/userFeedback.ts:34:14)
      at SourcePushDelete.errorMessageOnDeletion (src/commands/source/push/delete.ts:126:12)
      at errorMessageOnDeletion (src/commands/source/push/delete.ts:119:16)
          at async Promise.all (index 1)
      at SourcePushDelete.run (src/commands/source/push/delete.ts:82:7)
      at SourcePushDelete._run (../../../node_modules/@oclif/core/lib/command.js:81:22)
      at Config.runCommand (../../../node_modules/@oclif/core/lib/config/config.js:272:25)
      at Object.run (../../../node_modules/@oclif/test/lib/command.js:21:13)
      at Object.run (../../../node_modules/fancy-test/lib/base.js:68:25)

"
`;

exports[`source:push:delete returns an information message on deletion failure with older than 1`] = `
"  console.error
     â€º   Warning: Error while trying to delete document: older than 12345.
     â€º
     â€º   Status code: 412
     â€º   Error code: BAD_REQUEST
     â€º   Message: this is a bad request and you should feel bad

      32 |     throw error;
      33 |   } else {
    > 34 |     CliUx.ux.warn(error.message);
         |              ^
      35 |   }
      36 | };
      37 |

      at Object.warn (../../../node_modules/@oclif/core/lib/errors/index.js:59:13)
      at warn (src/lib/userFeedback.ts:34:14)
      at SourcePushDelete.errorMessageOnDeletion (src/commands/source/push/delete.ts:126:12)
      at SourcePushDelete.errorMessageOnDeletion [as doDeletionOlderThan] (src/commands/source/push/delete.ts:103:12)

"
`;

exports[`source:push:delete returns an information message on successful deletion with document uri 1`] = `
"The delete request for document: https://foo.com was accepted by the Push API.Status code: 999 this document is gone
    
"
`;

exports[`source:push:delete returns an information message on successful deletion with multiple document uri 1`] = `
"The delete request for document: https://foo.com was accepted by the Push API.Status code: 202 right trou
    
The delete request for document: https://foo.com/2 was accepted by the Push API.Status code: 202 right trou
    
"
`;

exports[`source:push:delete returns an information message on successful deletion with older than 1`] = `
"The delete request for document: older than 12345 was accepted by the Push API.Status code: 999 this document is gone
    
"
`;

exports[`source:push:delete should warn the user when he tries to delete too many items and stop there 1`] = `""`;

exports[`source:push:delete throws when incompatible flags for olderThan and documentUri are passed 1`] = `
"The following error occurred:
  [2m--delete=foo cannot also be provided when using --deleteOlderThan[22m
See more help with --help"
`;

exports[`source:push:delete throws when no flags are specified 1`] = `"You must minimally set the \`delete\` or the \`deleteOlderThan\` flag. Use \`source:push:delete --help\` to get more information."`;
